### PR TITLE
CXX-2074 implement mongohouse tests

### PR DESCRIPTION
**Pull request by samantharitter**
In order to support these tests, I added a few new methods to the bsoncxx::array::view class, along with tests.  I chose to implement them on the class itself rather than as a test method because I thought they would be generally useful methods for users of the driver.  These changes are in their own commit.  

I also made changes to allow tests to iterate over events in the apm checker, improve some error messaging around unset elements, and fixed a small bug around unset elements.

The mongohouse tests will run on evergreen in their own new linux buildvariant, where the mongohouse tests are the only task that runs.

### CODE REVIEW

File: src/bsoncxx/array/view.cpp
Comment (bazile-clyde) [inline src/bsoncxx/array/view.cpp:152]: We could also overload `operator[]` to be consistent with the other find/operator methods for indices. So just:
```
element view::operator[](bsoncxx::types::bson_value::view v) const {
    return *(this->find_value(v));
}
```
would allow for boolean testing.

File: src/bsoncxx/array/view.hpp
Comment (bazile-clyde) [inline src/bsoncxx/array/view.hpp:34]: Is this used somewhere?

**Review by bazile-clyde — CHANGES_REQUESTED** [review]

File: src/mongocxx/test/spec/monitoring.cpp
Comment (kevinAlbs) [inline src/mongocxx/test/spec/monitoring.cpp:46]: Nice!

File: src/mongocxx/test/spec/monitoring.hh
Comment (kevinAlbs) [inline src/mongocxx/test/spec/monitoring.hh:48]: Oh good idea. This does make checking capture events more convenient.

File: src/bsoncxx/array/view.hpp
Comment (samantharitter) [reply src/bsoncxx/array/view.hpp:98]: Done, I'll change this commits to reference CXX-2125.

File: src/mongocxx/test/spec/monitoring.hh
Comment (samantharitter) [reply src/mongocxx/test/spec/monitoring.hh:48]: Thanks!  It would have been difficult to implement these tests without this.

File: src/mongocxx/test/spec/monitoring.cpp
Comment (samantharitter) [reply src/mongocxx/test/spec/monitoring.cpp:46]: Thanks!

File: src/bsoncxx/document/element.cpp
Comment (samantharitter) [reply src/bsoncxx/document/element.cpp:95]: Thanks!  That was a fun bit of debugging...

File: src/bsoncxx/document/element.cpp
Comment (samantharitter) [reply src/bsoncxx/document/element.cpp:63]: Glad you think so! :D

File: src/bsoncxx/array/view.cpp
Comment (samantharitter) [reply src/bsoncxx/array/view.cpp:143]: I thought that might be confusing to users because the two methods work so differently.  The existing find() method looks up an element at the given index--which is not what I expected it to do when I first looked at this class!

Really, I think it would be more clear if the existing `array::find()` method were renamed to `array::at()`, to match the API of `std::array`.  Then we would be free to repurpose `array::find()` to act like`array::find_value()`, which would then also behave more like `std::find` and maybe be more clear to users.

If you and @kevinAlbs agree with all that, then here's what I think we could do:
- rename the existing `array::find()` to `array::at()` and deprecate `array::find()`
- implement `array::find_value()` as shown here and maybe add `array::find_value_if()`
- open a JIRA ticket to deprecate `array::find_value()` and rename it to `array::find()` in the next release (same for `find_value_if`)

Comment (samantharitter) [reply src/bsoncxx/array/view.cpp:143]: That's true, I could have implemented this that way!  But unless you have a strong preference for using `std::find_if`, I'll just leave it as-is, as they are essentially the same

File: src/bsoncxx/array/view.cpp
Comment (samantharitter) [reply src/bsoncxx/array/view.cpp:152]: I also don't think we should overload `operator[]`, I think we should try to keep methods in agreement with the `std::array` API as much as possible.  For `std::array`, `operator[]` functions as our existing one does, by looking up the element at the given index.

File: src/bsoncxx/array/view.hpp
Comment (samantharitter) [reply src/bsoncxx/array/view.hpp:34]: This forward declaration is necessary in order to make the `find_value` declaration below compile.

File: src/mongocxx/test/spec/mongohouse.cpp
Comment (samantharitter) [reply src/mongocxx/test/spec/mongohouse.cpp:88]: Ah whoops, will remove

File: src/bsoncxx/test/array.cpp
Comment (samantharitter) [reply src/bsoncxx/test/array.cpp:20]: Nope, good call!  I will remove it.

File: src/bsoncxx/array/view.cpp
Comment (kevinAlbs) [reply src/bsoncxx/array/view.cpp:143]: > rename the existing array::find() to array::at() and deprecate array::find()

> open a JIRA ticket to deprecate array::find_value() and rename it to array::find() in the next release (same for find_value_if)


I agree with `at` being a better name than `find`. I am in favor of adding `at`.

But I am opposed to deprecating and repurposing `find`.

Given that you can construct a `bson_value` from an int32, that might silently change behavior when a user upgrades.

Suppose a user has this code, intending to look up element at position 3:
```
auto iter = arr.find(3);
```
Then they upgrade to C++ 3.8, and we've change `find` to mean "find by searching bson_value", I think this would still compile, but have a different meaning.

> implement array::find_value() as shown here and maybe add array::find_value_if()

`find_value` seems like a clear enough name to me. As does `search`.

If we want to be consistent with `std::array`, it does not provide `find` or `find_if`. Instead, users use `std::find` and `std::find_if` with iterators. We could take the same approach. If we made a `bsoncxx::element` comparable to `bsoncxx::bson_value`, I think users could use `std::find` and `std::find_if`.

That being said `arr.find_value("test")` is more concise than `std::find (arr.begin(), arr.end(), "test"))`. `find_value` and `has_value` seem like reasonable conveniences. But I think we should draw the line somewhere too.

**Review by kevinAlbs — COMMENTED** [review]
Changes LGTM pending discussion of `has_value` and `find_value`.

File: src/bsoncxx/array/view.cpp
Comment (samantharitter) [reply src/bsoncxx/array/view.cpp:143]: Perhaps we should rename `find()` to `at()` (in a separate ticket) and leave this as `find_value()` then?

Comment (kevinAlbs) [reply src/bsoncxx/array/view.cpp:143]: SGTM if @bazile-clyde agrees.

Comment (bazile-clyde) [reply src/bsoncxx/array/view.cpp:143]: > I thought that might be confusing to users because the two methods work so differently. The existing find() method looks up an element at the given index--which is not what I expected it to do when I first looked at this class!

Ah, that makes sense. It’d be really confusing to have a subtle difference between finding by index or value.

> Really, I think it would be more clear if the existing array::find() method were renamed to array::at(), to match the API of std::array.

I think so too. We could (and probably should for consistency with `std::array`) also add bounds checking for `at`. So `throw` instead of returning the end iterator. Especially since we already have a non-throwing "find" with `operator[]`.

> Then we would be free to repurpose array::find() to act like array::find_value(), which would then also behave more like std::find and maybe be more clear to users.

That being said, I’m not sure why we need `find` or `find_value` though. Sequence containers in the STL don’t have `find` methods and instead rely on `std::find` to look for specific elements by value. The `find` methods for associative containers are used to look up elements by keys, not values. 

As is, `find_value` behaves identically to `std::find`. The only difference is how we compare the elements. Why not just create an `operator==` that compares array elements to values? Then users can use `std::find` without a custom compare function, which is essentially what `find_value` is abstracting away.

File: src/bsoncxx/array/view.cpp
Comment (bazile-clyde) [reply src/bsoncxx/array/view.cpp:152]: That's a good, point. Resolved.

File: src/mongocxx/test/spec/mongohouse.cpp
Comment (bazile-clyde) [reply src/mongocxx/test/spec/mongohouse.cpp:88]: Was the removal pushed?

File: src/mongocxx/test/spec/mongohouse.cpp
Comment (bazile-clyde) [inline src/mongocxx/test/spec/mongohouse.cpp:38]: Since this code is identical to `crud.cpp` maybe we should consolidate both chunks into one method and/or use some of the functions in `util.cpp` like the other test runners. It'll be a lot less code to update when we inevitably have to refactor this.

File: src/bsoncxx/array/view.cpp
Comment (samantharitter) [reply src/bsoncxx/array/view.cpp:143]: I think that adding `find_value` is a nice shortcut, and a little bit nicer to use than `std::find` for an operation that I imagine being somewhat common for users.  Especially considering that it might be a little confusing to users that `std::find` would compare `element` and `value` and not two things of the same type.  But, it does differ from the API of an STL container, and if you and @kevinAlbs agree that it would be better to encourage users to use `std::find` for element lookup, I'll change the test to use `std::find` and add public `element` to `value` comparison functions instead.

I've opened a separate ticket to rename `find()` to `at()`.

Comment (kevinAlbs) [reply src/bsoncxx/array/view.cpp:143]: I discussed this with @bazile-clyde. Though I do see the convenience of `find_value`, I am in favor of removing `find_value` in and adding an `operator==` overload for `bson_value` and `element` for consistency with STL.

Comment (samantharitter) [reply src/bsoncxx/array/view.cpp:143]: Fair enough! Done.

File: src/mongocxx/test/spec/mongohouse.cpp
Comment (samantharitter) [reply src/mongocxx/test/spec/mongohouse.cpp]: Yes, hopefully this is fixed in my most recent commit.  I believe it has to do with the curly bracket initialization, I once found the formal bug report and could find it again if you are interested.

File: src/mongocxx/test/spec/mongohouse.cpp
Comment (bazile-clyde) [reply src/mongocxx/test/spec/mongohouse.cpp:38]: Following up on this.

Comment (samantharitter) [reply src/mongocxx/test/spec/mongohouse.cpp:38]: Ah whoops, missed this comment.  Will refactor.

**Review by bazile-clyde — APPROVED** [review]
LGTM, pending fixes for test failures.

File: src/mongocxx/test/CMakeLists.txt
Comment (kevinAlbs) [inline src/mongocxx/test/CMakeLists.txt:209]: Why is this only on run on Linux? https://github.com/mongodb-labs/drivers-evergreen-tools/tree/master/.evergreen/atlas_data_lake accounts for running in Windows.

I am ok with limiting to Linux now. But I think it is worth testing connectivity on Windows / macOS since TLS goes through a different code path. But we could do this in C driver tests instead, since the C++ driver does not alter TLS / connection behavior.

**Review by kevinAlbs — COMMENTED** [review]
Code changes look good!

I have a hypothesis on the test failures.

In single threaded mode, a `mongocxx::client` will exit server selection if it is unable to connect after attempting one time.
https://github.com/mongodb/specifications/blob/master/source/server-selection/server-selection.rst#serverselectiontryonce

mongohouse is started in the background. If the C++ driver tests start running before mongohouse starts listening on sockets, then that could result in the immediate server selection error.

We could solve that by running mongohouse tests with the URI parameter `serverSelectionTryOnce=false` or by running some script before the tests that ensures mongohouse is listening for connections (e.g. loop and sleep for one second until the shell can connect).

File: src/mongocxx/test/CMakeLists.txt
Comment (samantharitter) [reply src/mongocxx/test/CMakeLists.txt:209]: That... is a great question.  I was using Java's implementation as a reference, and they only run the mongohouse test on an ubuntu-1804.  But now that I've gone looking, I can't find anywhere that states which/how many different buildvariants should be running these tests.  

In the scope document for ADL, there are these two goals:
- For each driver, run the test suite in Evergreen triggered off of commits to ADL's master branch
- For each driver, run the test suite in Evergreen on every commit to the driver.

Nothing specific about what type of environment the test suite should run in on Evergreen.  We are satisfying the second goal, at least.  I'm not sure about how the first one would be triggered, but I didn't see any of the other implementations checking for updates to the ADL master branch.

For now, maybe we can leave this as just linux?  It would be easy to add additional environments later, once there's more clear guidance about which environments should be tested.

**Comment by samantharitter** [comment]
Let's see if serverSelectionTryOnce=false fixes evergreen!  My patch build before this was green, so I think you're right about it being a flaky timing issue.

File: src/mongocxx/test/CMakeLists.txt
Comment (kevinAlbs) [reply src/mongocxx/test/CMakeLists.txt:209]: > Nothing specific about what type of environment the test suite should run in on Evergreen. We are satisfying the second goal, at least. I'm not sure about how the first one would be triggered, but I didn't see any of the other implementations checking for updates to the ADL master branch.

Oh good catch! I think that is done outside of code changes, by setting an [inter-project dependency](https://github.com/evergreen-ci/evergreen/wiki/Project-and-Distro-Settings#inter-project-dependencies) in [Evergreen project settings for cxx-driver](https://evergreen.mongodb.com/projects##cxx-driver). That can be done after this is merged.

For now, maybe we can leave this as just linux? It would be easy to add additional environments later, once there's more clear guidance about which environments should be tested.

File: src/mongocxx/test/spec/mongohouse.cpp
Comment (kevinAlbs) [inline src/mongocxx/test/spec/mongohouse.cpp:37]: Mind noting with a comment why `serverSelectionTryOnce=false` is set? The racy behavior of spawning mongohouse might not be obvious later.

**Review by kevinAlbs — APPROVED** [review]
Tests passed, LGTM with a comment!

File: src/mongocxx/test/spec/mongohouse.cpp
Comment (samantharitter) [reply src/mongocxx/test/spec/mongohouse.cpp:37]: Yes, definitely!

File: src/mongocxx/test/CMakeLists.txt
Comment (samantharitter) [reply src/mongocxx/test/CMakeLists.txt:209]: Will do!

**Comment by samantharitter** [comment]
Merged in!

**Pull request closed without merging**
